src/WebJobs.Script/Workers/Rpc/JobHostRpcWorkerChannelManager.cs

```diff
@@ -12,45 +13,86 @@ namespace Microsoft.Azure.WebJobs.Script.Workers.Rpc
     internal class JobHostRpcWorkerChannelManager : IJobHostRpcWorkerChannelManager
     {
         private readonly ILogger _logger;
-        private ConcurrentDictionary<string, IRpcWorkerChannel> _channels = new ConcurrentDictionary<string, IRpcWorkerChannel>();
+        private ConcurrentDictionary<string, IRpcWorkerChannelHolder> _channels = new ConcurrentDictionary<string, IRpcWorkerChannelHolder>();
 
         public JobHostRpcWorkerChannelManager(ILoggerFactory loggerFactory)
         {
             _logger = loggerFactory.CreateLogger<JobHostRpcWorkerChannelManager>();
         }
 
-        public void AddChannel(IRpcWorkerChannel channel)
+        public void AddChannel(IRpcWorkerChannel channel, string language)
         {
-            _channels.TryAdd(channel.Id, channel);
+            if (_channels.TryGetValue(language, out IRpcWorkerChannelHolder channels))
+            {
+                channels.TryAdd(channel.Id, channel);
+            }
+            else
+            {
+                _channels.TryAdd(language, new IRpcWorkerChannelHolder
+                {
+                    [channel.Id] = channel,
+                });
```

Comment:
I don't think this would be thread-safe -- for example if a bunch of threads all `AddChannel("node", ...)` at once, you could end up missing some of those as they all raced to call `_channels.TryAdd("node"...)`. I think we need a lock around this if it's being called multi-threaded anywhere.

Reply:
We were using `ConcurrentDictionary` before too. Currently we are creating single worker per language in sequence. What do you suggest here, should we have locks/waits here?

Reply:
Looks like the issue would be introduced in case of language worker handling and might not have existed before, added a synchronized block for this. Also added a test case to test the concurrency.